Refactor Vulkan context into its own files.  Use RAII.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42930 Vulkan command buffer and pool.
* #42709 Vulkan resource cache.
* #42642 Vulkan descriptor and descriptor layout cache.
* #42395 Vulkan pipeline and pipeline layout cache.
* #42786 Vulkan memory allocator.
* #42325 Vulkan (source and binary) shader and shader layout cache.
* #42394 Generic Vulkan object cache.
* #42503 Move torch/csrc/utils/hash.h to c10/util/hash.h.
* **#42273 Refactor Vulkan context into its own files.  Use RAII.**
* #43301 Search on system path for Vulkan headers and libraries as a last resort.

Differential Revision: [D23252335](https://our.internmc.facebook.com/intern/diff/D23252335)

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Context.cpp:
```diff
@@ -0,0 +1,281 @@
+        instance_layer_properties.data()));
+
+    constexpr std::array<const char*, 6> requested_instance_layers{
+        "VK_LAYER_GOOGLE_unique_objects",
```
There is some optimal validation layer order:
```
        "VK_LAYER_GOOGLE_threading",
        "VK_LAYER_LUNARG_parameter_validation",
        "VK_LAYER_LUNARG_object_tracker",
        "VK_LAYER_LUNARG_core_validation",
        "VK_LAYER_GOOGLE_unique_objects",
```
( https://www.lunarg.com/wp-content/uploads/2019/04/UberLayer_V3.pdf )

AshkanAliabadi replied:
Thanks.  I think even better is `VK_LAYER_KHRONOS_validation`:

https://vulkan.lunarg.com/doc/view/1.1.114.0/windows/khronos_validation_layer.html
https://vulkan.lunarg.com/doc/view/1.2.135.0/mac/validation_layers.html

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Context.cpp:
```diff
@@ -0,0 +1,281 @@
+
+Context::Context(const bool enable_validation_layers)
+    : instance_(create_instance(enable_validation_layers), &VK_DELETER(Instance)),
+      physical_device_(acquire_physical_device(instance())),
```
In this inits there are some places with using member functions to get `instance()` or member field instance_.
Could it be member field everywhere?

AshkanAliabadi replied:
You mean in the constructor or in general?  I tried to only use the accessor functions in the constructor for no particular reason, but yeah, only meant to consistently use the accessors and not mix with member variables - if that's the case I missed it, could you point it out?  Functions above and in the anonymous namespace are free functions by the way and access their input and output only through their parameters so if you refer to those that's why they don't use member functions.

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Context.cpp:
```diff
@@ -0,0 +1,281 @@
+        "Could not load vkCreateDebugReportCallbackEXT");
+
+    VkDebugReportCallbackEXT debug_report_callback{};
+    VK_CHECK(vkCreateDebugReportCallbackEXT(
```
debugReportCallback needs explicit destructor like:
```
    const auto func = (PFN_vkDestroyDebugReportCallbackEXT)vkGetInstanceProcAddr(
        instance_, "vkDestroyDebugReportCallbackEXT");
    if (func) {
      func(instance_, debugReportCallback, nullptr);
    }
```
( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkDestroyDebugReportCallbackEXT.html )

Here it is only as temporary variable in `create_instance()`
Does it need some special deleter function?

AshkanAliabadi replied:
Thanks for the catch.

AshkanAliabadi replied:
By the way, VK_EXT_validation_flags is deprecated.  At one point we should replace it with VK_EXT_debug_utils.  I'll do that later on.

AshkanAliabadi reviewed (commented):
Thanks for the review.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/api/Context.cpp:
```diff
@@ -0,0 +1,281 @@
+    0u,
+    nullptr,
+    0u,
+    nullptr,
```
By the way, instance layers (extensions) and device layers (extensions) are different.  I suspect this might be the reason why we have a segfault at shutdown.  I have not tested this hypothesis, but validation layers are instance layers (hence passing null here.)

Commit:
Update on "Refactor Vulkan context into its own files.  Use RAII."




[ghstack-poisoned]

Commit:
Update on "Refactor Vulkan context into its own files.  Use RAII."




[ghstack-poisoned]

Commit:
Update on "Refactor Vulkan context into its own files.  Use RAII."


Differential Revision: [D23252335](https://our.internmc.facebook.com/intern/diff/D23252335)

[ghstack-poisoned]

Commit:
Update on "Refactor Vulkan context into its own files.  Use RAII."


Differential Revision: [D23252335](https://our.internmc.facebook.com/intern/diff/D23252335)

[ghstack-poisoned]
